### PR TITLE
Update __init__.py

### PR DESCRIPTION
**Pull request by sresam89**
Fixing the "Problem parsing pin code string too many values to unpack (expected 4) "

### CODE REVIEW

**Comment by sresam89** [comment]
@pavoni or @vangorra  could you please approve the pull request

File: README.md
Comment (pavoni) [inline README.md:55]: The readme refers to `set_and_delete_door_code` - but the PR contains `delete_door_code` and `set_door_code`

File: examples/delete_door_code.py
Comment (pavoni) [inline examples/delete_door_code.py:48]: Perhaps better "\nLock command failed "

File: examples/set_door_code.py
Comment (pavoni) [inline examples/set_door_code.py:36]: `get_devices` same comment as above

File: examples/delete_door_code.py
Comment (pavoni) [inline examples/delete_door_code.py:33]: `get_devices` takes a second parameter to filter the device type - which might be slightly more elegant.

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:548]: I no longer have a live vera hub - which limits my ability to check changes. As such while I can see why you want to change the parameter structure -but I'm a little reluctant because testing the refactoring needed.

Even though it's less elegant I'd be happier for you to just add an extra call for the lock api - and leave the "set_service_value" method unchanged.

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:556]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:559]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:807]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:860]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:908]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:938]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:1024]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:1039]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:1241]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:1267]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:1295]: See above

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:1207]: If slots are integers - why do you want to change this to a dictionary?

**Review by pavoni — CHANGES_REQUESTED** [review]
Thanks for submitting this.

I no longer have a live vera hub myself. I still have it - but even if I fire it up it no longer has many devices attached - so it's not a great test.

For that reason I'd like to keep changes as simple as possible to minimise the risk for people who depend on this code.

As such I've suggested leaving the parameters of `set_service_value` unchanged - and just adding another  similar method to do what you need for the lock.

Not my usual preference - but in this case I'd prefer to limit change to the lock code.

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py:1207]: I did not request this change, it should have been from the tag I forked from
I made chagne to the line 1219
`_, _, pin, name = code_addrs[2:6]`  added `:6` to the list index

File: examples/delete_door_code.py
Comment (sresam89) [reply examples/delete_door_code.py:33]: I was following the existing code flow, minimizing no:of breaking changes

File: README.md
Comment (sresam89) [reply README.md:55]: I will update the readme on this

File: examples/delete_door_code.py
Comment (sresam89) [reply examples/delete_door_code.py:48]: I believe it would be  better to show the user what was the reason code is rather throwing a general message!!
There could be several reasons
1) Bad slot number
2) Slot already empty and nothing to clear
3) Lock is busy
4) Connection error (request timeout....)

Your thoughts?

File: examples/set_door_code.py
Comment (sresam89) [reply examples/set_door_code.py:36]: I was following the existing code flow, minimizing no:of breaking changes

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py:548]: I can do that....

File: examples/delete_door_code.py
Comment (pavoni) [reply examples/delete_door_code.py:48]: I was just suggesting you change the text so it was more obvious it was an error. Happy with including rest of detail as you have.

File: examples/set_door_code.py
Comment (pavoni) [reply examples/set_door_code.py:36]: My suggestion would improve it slightly and avoid the need to do the `isinstance` check.

File: pyvera/__init__.py
Comment (pavoni) [reply pyvera/__init__.py:1207]: I guess you may need to rebase. 

But you’re branch currently includes this change which isn’t in the library.

File: pyvera/__init__.py
Comment (pavoni) [reply pyvera/__init__.py:548]: Open minded.  Perhaps `call_service_with_params` or something like that?

File: examples/set_door_code.py
Comment (sresam89) [reply examples/set_door_code.py:36]: I could, but I would have to change the get_devices() method on all over the code, which could be a breaking change on hassio

Comment (pavoni) [reply examples/set_door_code.py:36]: Take a look at the code, there is an optional second parameter. You only need to change your call!

https://github.com/pavoni/pyvera/blob/941823e1a13c8ea2d4dd398287939bf0c6f970f6/pyvera/__init__.py#L196

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py:571]: used set_door_code_values

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py]: Addressed

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py]: Addressed

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py]: Addressed

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py]: Addressed

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py:1230]: Sorry My bad, I did change code to be dict over list. 
I am formatting this more elegantly, over output string being just a CSV list.

Previously it was
` codes.append((slot, name, pin))` I changed to be `codes[slot]={'name': name, 'pin': pin}`

There is no dependency on hassio for this, this will be a new feature request there

Comment (pavoni) [reply pyvera/__init__.py:1230]: You changed codes from an array to a dict.

please put it back.

Comment (sresam89) [reply pyvera/__init__.py:1230]: Is the Change from array to dict not acceptable? Is it breaking anything else?

**Comment by pavoni** [comment]
Just for clarity - you also need to fix the build error. 

Looks like your imports are not sorted correctly.

File: pyvera/__init__.py
Comment (pavoni) [reply pyvera/__init__.py:1230]: You already know my views.

Comment (sresam89) [reply pyvera/__init__.py:1230]: Sorry not getting emotions over comments.
I would like to hear it if it is not acceptable and understand the reasoning why
I would love to discuss/hear about the same on other chat/messenger/meeting platforms if necessary

Comment (pavoni) [reply pyvera/__init__.py:1230]: As you already know I’m trying to keep this code stable.

I don’t currently see any good reason for changing `codes` from an array to a dict.

Comment (sresam89) [reply pyvera/__init__.py:1230]: Having the output as dict helps a ton on the hassio side or other applications that will have to parse a CSV otherwise. 
Looks like that is not good enough of an argument for you

Comment (pavoni) [reply pyvera/__init__.py:1230]: It doesn’t make any sense to me.

As far as I can see you’re creating a dictionary with a key of `slot` which is an integer.

Each value of the `codes` dictionary is itself a dictionary - and I agree these make sense because the keys are ‘ name` and `pin`.

I expect if you just change back the initial assignment of `codes` you code will work because you’re assigning into it with an array assignment anyway (I agree the assignment is better than the append)

So honestly it just looks like an mistake.

Comment (sresam89) [reply pyvera/__init__.py:1230]: Hope you agree to correct this mistake someday. 

Probably another PR to fight for. 

Am reverting to original for now.

**Comment by sresam89** [comment]
Made all the changes requested (except for passing the deviceIsntance to lock) and pushed the code, let me know your comments

> Just for clarity - you also need to fix the build error.
> 
> Looks like your imports are not sorted correctly.

Could you help me with this
Am running this on a windows machine. Not able to use your build scripts

File: pyvera/__init__.py
Comment (pavoni) [reply pyvera/__init__.py]: It’s not a mistake on my part.

Another comment like this I will stop reviewing and close this PR.

File: examples/delete_door_code.py
Comment (pavoni) [reply examples/delete_door_code.py:33]: I think you still need to make this change.

File: pyvera/__init__.py
Comment (pavoni) [inline pyvera/__init__.py:552]: Please put back this blank line to maintain consistency with the rest of the code base.

Comment (sresam89) [reply pyvera/__init__.py]: Done

File: examples/set_door_code.py
Comment (sresam89) [reply examples/set_door_code.py]: Done

File: examples/delete_door_code.py
Comment (sresam89) [reply examples/delete_door_code.py]: Done

File: pyvera/__init__.py
Comment (sresam89) [reply pyvera/__init__.py]: addressed

**Comment by pavoni** [comment]
What’s currently failing is ‘black’. Even if you can’t run the build script, you should be able to run the individual utilities.

You can see what’s being run here: https://github.com/pavoni/pyvera/blob/master/scripts/build.sh

If you can’t track down the issues I’ll check out your branch and run them, but I won’t be able to do that until mid next week.

**Comment by pavoni** [comment]
You can read the build results here (although the black error isn’t very helpful!)

https://github.com/pavoni/pyvera/pull/151/checks?check_run_id=3615579670

**Comment by sresam89** [comment]
> What’s currently failing is `black`. Even if you can’t run the build script, you should be able to run the individual utilities.
> 
> You can see what’s being run here: https://github.com/pavoni/pyvera/blob/master/scripts/build.sh
> 
> If you can’t track down the issues I’ll check out your branch and run them, but I won’t be able to do that until mid next week.

Fixed black error wish

**Comment by pavoni** [comment]
You’ve still got black error on your changes to `__init__.py`

**Comment by sresam89** [comment]
> You’ve still got black errors on your changes to `__init__.py`

Not on my local machine @pavoni see the screenshot
image

Commit Id 0a69a9648af178d0e2d6e5a7b333bbee5eeffd97 checkout clean with black on my local machine

**Comment by pavoni** [comment]
Don’t know if this helps you get stuff running. https://dev.to/bowmanjd/getting-started-with-python-poetry-3ica

Perhaps double check which black version you have.

Otherwise I’ll run it locally next week and let you know the problem.

Also worth making sure flake8, pylint etc run cleanly.

**Comment by sresam89** [comment]
> Don’t know if this helps you get stuff running. https://dev.to/bowmanjd/getting-started-with-python-poetry-3ica
> 
> Perhaps double check which black version you have.
> 
> Otherwise I’ll run it locally next week and let you know the problem.
> 
> Also worth making sure flake8, pylint etc run cleanly.

i got hold a linux pc and ran the build script, and now lint complaint about the following, which does not makes sense to me.
Strangely the same error goes away if include the `isinstance` check you asked me to remove
image

**Comment by sresam89** [comment]
Adding a test commit to verify possible Lint Issues 
[https://github.com/python/mypy/issues/8763](url)

**Review by sresam89 — COMMENTED** [review]
Adding test commit to verify lint issues

**Comment by sresam89** [comment]
I was able to fix almost every error please review the changes, also added test cases

**Comment by pavoni** [comment]
The build is still failing saying 4 files would be reformatted.

I’ll run locally and post the results in the next few days if you’re unable to reproduce with the ‘build’ script locally.

**Review by sresam89 — COMMENTED** [review]
I was able to resolve almost every build error, and also added test cases please review

**Comment by pavoni** [comment]
You can see the failing build here!

https://github.com/pavoni/pyvera/pull/151/checks?check_run_id=3646361399

**Review by sresam89 — COMMENTED** [review]
Black shows only 3 files for me. updated them

**Comment by pavoni** [comment]
This is the change black made for me:-

```
diff --git a/pyvera/__init__.py b/pyvera/__init__.py
index dffc8bc..4e0a4f6 100755
--- a/pyvera/__init__.py
+++ b/pyvera/__init__.py
@@ -566,20 +566,13 @@ class VeraDevice:
         )

     def set_door_code_values(
-        self,
-        service_id: Union[str, Tuple[str, ...]],
-        operation: str,
-        parameter: dict,
+        self, service_id: Union[str, Tuple[str, ...]], operation: str, parameter: dict
     ) -> requests.Response:
         """Add or remove door code on the vera Lock.

         This will call the Vera api to change Lock code.
         """
-        payload = {
-            "id": "lu_action",
-            "action": operation,
-            "serviceId": service_id,
-        }
+        payload = {"id": "lu_action", "action": operation, "serviceId": service_id}
         for param in parameter:
             payload[param] = parameter[param]
         result = self.vera_request(**payload)
(END)
```

**Comment by sresam89** [comment]
> diff --git a/pyvera/__init__.py b/pyvera/__init__.py

Which version of black are you using?

**Review by sresam89 — COMMENTED** [review]
I think i figured the problem.
I was using the 21.9b0 and you were using some version of 19.x

**Comment by pavoni** [comment]
‘poetry’ should be taking care of that. You can see ‘19.3b0’ in the lock file.

if you’re comfortable - I’ll just try and commit an update of ‘__init__’ that works to your branch.

**Comment by pavoni** [comment]
Black still wants you to remove the final `,`

```
diff --git a/pyvera/__init__.py b/pyvera/__init__.py
index 648dbff..4e0a4f6 100755
--- a/pyvera/__init__.py
+++ b/pyvera/__init__.py
@@ -566,7 +566,7 @@ class VeraDevice:
         )

     def set_door_code_values(
-        self, service_id: Union[str, Tuple[str, ...]], operation: str, parameter: dict,
+        self, service_id: Union[str, Tuple[str, ...]], operation: str, parameter: dict
     ) -> requests.Response:
         """Add or remove door code on the vera Lock.

(END)
```

**Review by pavoni — APPROVED** [review]

**Comment by pavoni** [comment]
Apart from the final black change I think this is now GTG.

Happy to merge and release when the build is clean.

**Comment by sresam89** [comment]
> ‘poetry’ should be taking care of that. You can see ‘19.3b0’ in the lock file.
> 
> if you’re comfortable - I’ll just try and commit an update of ‘**init**’ that works to your branch.

poetry is not doing that with respect to black (I had to install it manually)
```
(.venv) sreeram@ubuntu-sam:~/hassio/pyvera$ poetry install
Installing dependencies from lock file                                 

No dependencies to install or update

Installing the current project: pyvera (0.3.14)
(.venv) sreeram@ubuntu-sam:~/hassio/pyvera$ pip list | grep b*
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.

(.venv) sreeram@ubuntu-sam:~/hassio/pyvera$ which black
/home/sreeram/.local/bin/black
(.venv) sreeram@ubuntu-sam:~/hassio/pyvera$ black --version
black, version 19.10b0
```

I uninstalled black at this point to show you
```
(.venv) sreeram@ubuntu-sam:~/hassio/pyvera$ poetry install
Installing dependencies from lock file

No dependencies to install or update

Installing the current project: pyvera (0.3.14)
(.venv) sreeram@ubuntu-sam:~/hassio/pyvera$ black --version
bash: /home/sreeram/.local/bin/black: No such file or directory
(.venv) sreeram@ubuntu-sam:~/hassio/pyvera$ 
```

**Review by sresam89 — COMMENTED** [review]
updating black changes (hopefully the last time)
you can make anymore changes, am ok with that

**Comment by sresam89** [comment]
How long does this take to update on https://pypi.org/project/pyvera/ ??

**Comment by pavoni** [comment]
I need to do a release - which I’ll do later today.

The update is then almost instant.

**Comment by pavoni** [comment]
Now released as 0.3.14.

Thanks for the contribution.

**Comment by sresam89** [comment]
I know we had a conversation on this already. Is it worth re-visiting the "pincode" being dict versus list ?
I totally understand your view on maintaining the stable code base, just trying an attempt to make it better.

**Comment by pavoni** [comment]
For items which are indexed only by an integer - the best data structure is really an array rather than a dict.

But unless there are a lot of pins - and a decent amount of looking up going on I think the advantages of changing to another data structure are small - and offset by the fact that we might break existing code using the API.

The library is't being actively developed - so I think t's wise to keep non critical changes to a minimum.

**Pull request merged**
